ospfd/ospf_asbr.c

```diff
@@ -329,3 +331,885 @@ void ospf_redistribute_withdraw(struct ospf *ospf, uint8_t type,
+		}
+
+		asel = (struct as_external_lsa *)lsa->data;
+		tag = (unsigned long)ntohl(asel->e[0].route_tag);
+
+		/* If tag modified , then re-originate the route
+		 * with modified tag details.
+		 */
+		if (tag != ei_aggr.tag) {
+			if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+				zlog_debug(
+					"%s: Route tag changed(old:%d new:%d,So refresh the summary route.(%pI4/%d)",
+					__func__, tag, ei_aggr.tag,
+					&aggr->p.prefix, aggr->p.prefixlen);
+
+			ospf_external_lsa_refresh(ospf, lsa, &ei_aggr,
+						  LSA_REFRESH_FORCE, 1);
+		}
+		return lsa;
+	}
+
+	if (lsa && IS_LSA_MAXAGE(lsa)) {
+		/* This is special case.
+		 * If a summary route need to be originated but where
+		 * summary route already exist in lsdb with maxage, then
+		 * it need to be refreshed.
+		 */
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: LSA is in MAX-AGE so refreshing LSA(%pI4/%d)",
+				__PRETTY_FUNCTION__, &aggr->p.prefix,
+				aggr->p.prefixlen);
+
+		ospf_external_lsa_refresh(ospf, lsa, &ei_aggr,
+					  LSA_REFRESH_FORCE, 1);
+		SET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+		return lsa;
+	}
+
+	/* If the external route prefix same as aggregate route
+	 * and if external route is already originated as TYPE-5
+	 * then it need to be refreshed and originate bit should
+	 * be set.
+	 */
+	if (lsa && prefix_same((struct prefix *)&ei_aggr.p,
+			       (struct prefix *)&ei->p)) {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug(
+				"%s: External route prefix is same as aggr so refreshing LSA(%pI4/%d)",
+				__PRETTY_FUNCTION__, &aggr->p.prefix,
+				aggr->p.prefixlen);
+		ospf_external_lsa_refresh(ospf, lsa, &ei_aggr,
+					  LSA_REFRESH_FORCE, 1);
+		SET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+		return lsa;
+	}
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("%s: Originate Summary route(%pI4/%d)", __func__,
+			   &aggr->p.prefix, aggr->p.prefixlen);
+
+	/* Originate summary LSA */
+	lsa = ospf_external_lsa_originate(ospf, &ei_aggr);
+	if (lsa) {
+		if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+			zlog_debug("%s: Set the origination bit for aggregator",
+				   __func__);
+		SET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+	}
+
+	return lsa;
+}
+void ospf_unset_all_aggr_flag(struct ospf *ospf)
+{
+	struct route_node *rn = NULL;
+
+	if (IS_DEBUG_OSPF(lsa, EXTNL_LSA_AGGR))
+		zlog_debug("Unset the origination bit for all aggregator");
+
+	for (rn = route_top(ospf->rt_aggr_tbl); rn; rn = route_next(rn)) {
+		if (!rn->info)
+			continue;
+
+		struct ospf_external_aggr_rt *aggr = rn->info;
+
+		UNSET_FLAG(aggr->flags, OSPF_EXTERNAL_AGGRT_ORIGINATED);
+	}
+}
+
+static void ospf_delete_all_marked_aggregators(struct ospf *ospf)
+{
+	struct route_node *rn = NULL;
+
+	/* Loop through all the aggregators, Delete all aggregators
+	 * which are marked as DELETE. Set action to NONE for remaining
+	 * aggregators
+	 */
+	for (rn = route_top(ospf->rt_aggr_tbl); rn; rn = route_next(rn))
+		if (rn->info) {
```

Comment:
if (!rn->info) continue?  Let's stay consistent in our approach from function to function?

Reply:
done.